src/main/java/sparksoniq/jsoniq/item/FunctionItem.java

```diff
@@ -135,7 +132,8 @@ public void write(Kryo kryo, Output output) {
         kryo.writeObject(output, this.identifier);
         kryo.writeObject(output, this.parameterNames);
         kryo.writeObject(output, this.signature);
-        kryo.writeObject(output, this.bodyExpression);
+        kryo.writeObject(output, this.bodyIterator.getClass());
+        kryo.writeObject(output, this.bodyIterator);
```

Comment:
You could try to serialize bodyIterator to byte[] using java.io.Serializable, and then write this as a byte array to kryo. That way you circumvent the (possibly untested and broken) kryo serialization of iterators. This is what I suggested before.

Reply:
If you do, however, want to use kryo for iterators, I think that writeClassAndObject may be what you are looking for here (and possibly the cause of error).